File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int dfs(TreeNode* root,int & maxdia){
        if(root == nullptr)
            return 0;
        int l=dfs(root->left,maxdia);
        int r=dfs(root->right,maxdia);
        int currdia = l+r;
        maxdia = max(maxdia,currdia);
        return 1+max(l,r);
    }

public:
    int diameterOfBinaryTree(TreeNode* root) {
        int maxdia =0;
        dfs(root,maxdia);
        return maxdia;

    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    private int maxdia = 0;

    private int dfs(TreeNode root) {

        if (root == null)
            return 0;

        int l = dfs(root.left);
        int r = dfs(root.right);

        int currdia = l + r;

        maxdia = Math.max(maxdia, currdia);

        return 1 + Math.max(l, r);
    }

    public int diameterOfBinaryTree(TreeNode root) {

        dfs(root);

        return maxdia;
    }
}
